include_attribute "hadoop"

#default[:saasfee][:resolution]                       = "1024x768x32"
default[:saasfee][:release]                           = "false"
default[:saasfee][:user]                              = "hiway"
default[:saasfee][:home]                              = "/home/#{node[:saasfee][:user]}"
default[:saasfee][:workflows]                         = "#{node[:saasfee][:home]}"
default[:saasfee][:data]                              = "#{node[:saasfee][:home]}/data"
default[:saasfee][:software][:dir]                    = "#{node[:saasfee][:home]}/software"

default[:saasfee][:hiway][:release][:version]         = "1.0.0-beta"
default[:saasfee][:hiway][:release][:targz]           = "hiway-dist-#{node[:saasfee][:hiway][:release][:version]}.tar.gz"
default[:saasfee][:hiway][:release][:url]             = "https://github.com/marcbux/Hi-WAY/releases/download/#{node[:saasfee][:hiway][:release][:version]}/#{node[:saasfee][:hiway][:release][:targz]}"
default[:saasfee][:hiway][:github_url]                = "https://github.com/marcbux/Hi-WAY.git"
default[:saasfee][:hiway][:home]                      = "#{node[:saasfee][:software][:dir]}/hiway"
default[:saasfee][:hiway][:hdfs][:basedir]            = "/user/#{node[:saasfee][:user]}/"
default[:saasfee][:hiway][:am][:memory_mb]            = 512
default[:saasfee][:hiway][:am][:vcores]               = 1
default[:saasfee][:hiway][:worker][:memory_mb]        = 1024
default[:saasfee][:hiway][:worker][:vcores]           = 1
default[:saasfee][:hiway][:scheduler]                 = "placementAware"

default[:saasfee][:cuneiform][:release][:version]     = "2.0.0-beta"
default[:saasfee][:cuneiform][:release][:targz]       = "cuneiform-dist-#{node[:saasfee][:cuneiform][:release][:version]}.tar.gz"
default[:saasfee][:cuneiform][:release][:url]         = "https://github.com/joergen7/cuneiform/releases/download/#{node[:saasfee][:cuneiform][:release][:version]}/#{node[:saasfee][:cuneiform][:release][:targz]}"
default[:saasfee][:cuneiform][:github_url]            = "https://github.com/joergen7/cuneiform.git"
default[:saasfee][:cuneiform][:home]                  = "#{node[:saasfee][:software][:dir]}/cuneiform"
default[:saasfee][:cuneiform][:r_packages]            = "#{node[:saasfee][:software][:dir]}/r_packages"
default[:saasfee][:cuneiform][:cache]                 = "/tmp/cf-cache"

default[:saasfee][:galaxy][:repository]               = "https://bitbucket.org/galaxy/galaxy-dist/"
default[:saasfee][:galaxy][:home]                     = "#{node[:saasfee][:software][:dir]}/galaxy"
default[:saasfee][:galaxy][:master_api_key]           = "reverse"
# name of at least four lower-case characters, numbers, or the "-" character
default[:saasfee][:galaxy][:user][:name]              = "#{node[:saasfee][:user]}"
# password of at least six characters
default[:saasfee][:galaxy][:user][:password]          = "#{node[:saasfee][:galaxy][:master_api_key]}"
default[:saasfee][:galaxy][:user][:email]             = "hiway@hiway.com"

default[:saasfee][:helloworld][:workflow]             = "helloworld.cf"

default[:saasfee][:wordcount][:workflow]              = "wordcount.cf"
default[:saasfee][:wordcount][:input][:url]           = "http://stateoftheunion.onetwothree.net/texts/stateoftheunion1790-2014.txt.zip"
default[:saasfee][:wordcount][:input][:zip]           = "stateoftheunion1790-2014.zip"
default[:saasfee][:wordcount][:input][:txt]           = "stateoftheunion1790-2014.txt"

default[:saasfee][:variantcall][:workflow]              = "variantcall.cf"
default[:saasfee][:variantcall][:threads]               = "#{node[:saasfee][:hiway][:worker][:vcores]}"
default[:saasfee][:variantcall][:memory_mb]             = "#{node[:saasfee][:hiway][:worker][:memory_mb]}"

default[:saasfee][:variantcall][:reads][:sample_id]     = "HG02025"
default[:saasfee][:variantcall][:reads][:run_ids]       = ["SRR359188", "SRR359195"]
default[:saasfee][:variantcall][:reads][:url_base]      = "ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/#{node[:saasfee][:variantcall][:reads][:sample_id]}/sequence_read"

default[:saasfee][:variantcall][:reference][:id]          = "hg38" 
default[:saasfee][:variantcall][:reference][:chromosomes] = ["chr22", "chrY"]
default[:saasfee][:variantcall][:reference][:url_base]    = "ftp://hgdownload.soe.ucsc.edu/apache/htdocs/goldenPath/#{node[:saasfee][:variantcall][:reference][:id]}/chromosomes"

default[:saasfee][:variantcall][:annovardb][:directory] = "annodb"
default[:saasfee][:variantcall][:annovardb][:file]      = "hg38db.tar"
default[:saasfee][:variantcall][:bowtie2][:version]   = "2.2.5"
default[:saasfee][:variantcall][:bowtie2][:zip]       = "bowtie2-#{node[:saasfee][:variantcall][:bowtie2][:version]}-linux-x86_64.zip"
default[:saasfee][:variantcall][:bowtie2][:home]      = "#{node[:saasfee][:software][:dir]}/bowtie2-#{node[:saasfee][:variantcall][:bowtie2][:version]}"
default[:saasfee][:variantcall][:bowtie2][:url]       = "http://downloads.sourceforge.net/project/bowtie-bio/bowtie2/#{node[:saasfee][:variantcall][:bowtie2][:version]}/#{node[:saasfee][:variantcall][:bowtie2][:zip]}"
default[:saasfee][:variantcall][:bwa][:version]       = "0.7.12"
default[:saasfee][:variantcall][:bwa][:tarbz2]        = "bwa-#{node[:saasfee][:variantcall][:bwa][:version]}.tar.bz2"
default[:saasfee][:variantcall][:bwa][:home]          = "#{node[:saasfee][:software][:dir]}/bwa-#{node[:saasfee][:variantcall][:bwa][:version]}"
default[:saasfee][:variantcall][:bwa][:url]           = "http://downloads.sourceforge.net/project/bio-bwa/#{node[:saasfee][:variantcall][:bwa][:tarbz2]}"
default[:saasfee][:variantcall][:samtools][:version]  = "1.2"
default[:saasfee][:variantcall][:samtools][:tarbz2]   = "samtools-#{node[:saasfee][:variantcall][:samtools][:version]}.tar.bz2"
default[:saasfee][:variantcall][:samtools][:home]     = "#{node[:saasfee][:software][:dir]}/samtools-#{node[:saasfee][:variantcall][:samtools][:version]}"
default[:saasfee][:variantcall][:samtools][:url]      = "http://download.sourceforge.net/project/samtools/samtools/#{node[:saasfee][:variantcall][:samtools][:version]}/#{node[:saasfee][:variantcall][:samtools][:tarbz2]}"
default[:saasfee][:variantcall][:varscan][:version]   = "2.3.7"
default[:saasfee][:variantcall][:varscan][:jar]       = "VarScan.v#{node[:saasfee][:variantcall][:varscan][:version]}.jar"
default[:saasfee][:variantcall][:varscan][:home]      = "#{node[:saasfee][:software][:dir]}/varscan-#{node[:saasfee][:variantcall][:samtools][:version]}"
default[:saasfee][:variantcall][:varscan][:url]       = "http://downloads.sourceforge.net/project/varscan/#{node[:saasfee][:variantcall][:varscan][:jar]}"
default[:saasfee][:variantcall][:annovar][:targz]     = "annovar.latest.tar.gz"
default[:saasfee][:variantcall][:annovar][:home]      = "#{node[:saasfee][:software][:dir]}/annovar"
default[:saasfee][:variantcall][:annovar][:url]       = "http://www.openbioinformatics.org/annovar/download/0wgxR2rIVP/#{node[:saasfee][:variantcall][:annovar][:targz]}"
default[:saasfee][:variantcall][:fastqc][:version]    = "0.11.4"
default[:saasfee][:variantcall][:fastqc][:zip]        = "fastqc_v#{node[:saasfee][:variantcall][:fastqc][:version]}.zip"
default[:saasfee][:variantcall][:fastqc][:home]       = "#{node[:saasfee][:software][:dir]}/FastQC"
default[:saasfee][:variantcall][:fastqc][:url]        = "http://www.bioinformatics.babraham.ac.uk/projects/fastqc/#{node[:saasfee][:variantcall][:fastqc][:zip]}"

default[:saasfee][:variantcall][:scale][:workflow]    = "variantcall-scale.cf"
default[:saasfee][:variantcall][:scale][:fa]          = "#{node[:saasfee][:data]}/#{node[:saasfee][:variantcall][:reference][:id]}/#{node[:saasfee][:variantcall][:reference][:id]}.fa"
default[:saasfee][:variantcall][:scale][:index]       = "#{node[:saasfee][:data]}/#{node[:saasfee][:variantcall][:reference][:id]}/hg19"
default[:saasfee][:variantcall][:scale][:db]          = "#{node[:saasfee][:data]}/#{node[:saasfee][:variantcall][:reference][:id]}/db/"
default[:saasfee][:variantcall][:scale][:gz]          = { "HG02025" => ["SRR359188", "SRR359195"] }

default[:saasfee][:montage_synthetic][:workflow]       = "Montage_25.xml"
default[:saasfee][:montage_synthetic][:url]            = "https://confluence.pegasus.isi.edu/download/attachments/2490624/#{node[:saasfee][:montage_synthetic][:workflow]}"

default[:saasfee][:montage_m17_4][:workflow]           = "montage_m17_4.dax"
default[:saasfee][:montage_m17_4][:montage][:version]  = "3.3"
default[:saasfee][:montage_m17_4][:montage][:targz]    = "Montage_v#{node[:saasfee][:montage_m17_4][:montage][:version]}.tar.gz"
default[:saasfee][:montage_m17_4][:montage][:home]     = "#{node[:saasfee][:software][:dir]}/Montage_v#{node[:saasfee][:montage_m17_4][:montage][:version]}"
default[:saasfee][:montage_m17_4][:montage][:url]      = "http://montage.ipac.caltech.edu/download/#{node[:saasfee][:montage_m17_4][:montage][:targz]}"

default[:saasfee][:galaxy101][:workflow]               = "galaxy101.ga"
default[:saasfee][:galaxy101][:exons][:bed]            = "Exons.bed"
default[:saasfee][:galaxy101][:exons][:targz]          = "#{node[:saasfee][:galaxy101][:exons][:bed]}.tar.gz"
default[:saasfee][:galaxy101][:snps][:bed]             = "SNPs.bed"
default[:saasfee][:galaxy101][:snps][:targz]           = "#{node[:saasfee][:galaxy101][:snps][:bed]}.tar.gz"
default[:saasfee][:galaxy101][:join][:name]            = "join"
default[:saasfee][:galaxy101][:join][:revision]        = "de21bdbb8d28"

default[:saasfee][:RNAseq][:workflow]                            = "RNAseq.ga"
default[:saasfee][:RNAseq][:ref_annotation][:gtf]                = "mm9_ref_annotation.gtf"
default[:saasfee][:RNAseq][:ref_annotation][:targz]              = "#{node[:saasfee][:RNAseq][:ref_annotation][:gtf]}.tar.gz"
default[:saasfee][:RNAseq][:sratoolkit][:version]                = "2.4.5-2"
default[:saasfee][:RNAseq][:sratoolkit][:targz]                  = "sratoolkit.#{node[:saasfee][:RNAseq][:sratoolkit][:version]}-ubuntu64.tar.gz"
default[:saasfee][:RNAseq][:sratoolkit][:home]                   = "#{node[:saasfee][:software][:dir]}/sratoolkit.#{node[:saasfee][:RNAseq][:sratoolkit][:version]}-ubuntu64"
default[:saasfee][:RNAseq][:sratoolkit][:url]                    = "http://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/#{node[:saasfee][:RNAseq][:sratoolkit][:version]}/#{node[:saasfee][:RNAseq][:sratoolkit][:targz]}"
default[:saasfee][:RNAseq][:input1][:replicate1][:accession]    = "SRR1632911"
default[:saasfee][:RNAseq][:input1][:replicate1][:fastq]        = "GSM1533014_MD_O1_WT_Colon.fastqsanger"
default[:saasfee][:RNAseq][:input1][:replicate2][:accession]    = "SRR1632912"
default[:saasfee][:RNAseq][:input1][:replicate2][:fastq]        = "GSM1533014_MD_O2_WT_Colon.fastqsanger"
default[:saasfee][:RNAseq][:input1][:replicate3][:accession]    = "SRR1632913"
default[:saasfee][:RNAseq][:input1][:replicate3][:fastq]        = "GSM1533014_MD_O3_WT_Colon.fastqsanger"
default[:saasfee][:RNAseq][:input2][:replicate1][:accession]    = "SRR1632942"
default[:saasfee][:RNAseq][:input2][:replicate1][:fastq]        = "GSM1533014_MD_Y1_WT_Colon.fastqsanger"
default[:saasfee][:RNAseq][:input2][:replicate2][:accession]    = "SRR1632943"
default[:saasfee][:RNAseq][:input2][:replicate2][:fastq]        = "GSM1533014_MD_Y2_WT_Colon.fastqsanger"
default[:saasfee][:RNAseq][:input2][:replicate3][:accession]    = "SRR1632944"
default[:saasfee][:RNAseq][:input2][:replicate3][:fastq]        = "GSM1533014_MD_Y3_WT_Colon.fastqsanger"
default[:saasfee][:RNAseq][:fastq_trimmer_by_quality][:name]     = "fastq_trimmer_by_quality"
default[:saasfee][:RNAseq][:fastq_trimmer_by_quality][:revision] = "1cdcaf5fc1da"
default[:saasfee][:RNAseq][:fastqc][:name]                       = "fastqc"
default[:saasfee][:RNAseq][:fastqc][:revision]                   = "e28c965eeed4"
default[:saasfee][:RNAseq][:fastx_clipper][:name]                = "fastx_clipper"
default[:saasfee][:RNAseq][:fastx_clipper][:revision]            = "8192b4014977"
default[:saasfee][:RNAseq][:tophat2][:name]                      = "tophat2"
default[:saasfee][:RNAseq][:tophat2][:revision]                  = "ae06af1118dc"
default[:saasfee][:RNAseq][:picard][:name]                       = "picard"
default[:saasfee][:RNAseq][:picard][:revision]                   = "ab1f60c26526"
default[:saasfee][:RNAseq][:cufflinks][:name]                    = "cufflinks"
default[:saasfee][:RNAseq][:cufflinks][:revision]                = "9aab29e159a7"
default[:saasfee][:RNAseq][:cuffmerge][:name]                    = "cuffmerge"
default[:saasfee][:RNAseq][:cuffmerge][:revision]                = "424d49834830"
default[:saasfee][:RNAseq][:cuffcompare][:name]                  = "cuffcompare"
default[:saasfee][:RNAseq][:cuffcompare][:revision]              = "67695d7ff787"
default[:saasfee][:RNAseq][:cuffdiff][:name]                     = "cuffdiff"
default[:saasfee][:RNAseq][:cuffdiff][:revision]                 = "604fa75232a2"
default[:saasfee][:RNAseq][:column_maker][:name]                 = "column_maker"
default[:saasfee][:RNAseq][:column_maker][:revision]             = "08a01b2ce4cd"
